ESQL: Date math for negatives

alex-spies opened the pull request:
Currently, negating periods/durations works in date math expressions, like `now() + (-1hour)` or `now() + (-1year)`, but does not once the duration/period is not a literal.

- Enable expressions like `now() + (-(1hour + 1minute))`.
- Add `DateMathException` for cases when this e.g. overflows and we need to send back a 400 error to the user.

alex-spies committed:
Enable folding negation of date math expressions

alex-spies committed:
Use more realistic durations in function tests

Nanosecond literals are currently not supported, use millis.

alex-spies committed:
Make Neg test cases more precise

alex-spies committed:
Factor out IllegalTemporalValueException

alex-spies committed:
Handle overflows when negating consistently

Other date arithmetic throws an exception rather than returning null + a
warning. Make period/duration negation consistent with that.

alex-spies committed:
Update REST tests, add negation case

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Hi @alex-spies, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

alex-spies committed:
Update docs/changelog/99711.yaml

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateMathException.java:
```diff
@@ -0,0 +1,33 @@
+ * Even though this is a pretty specific exception, we cannot use e.g. a anonymous class that inherits from {@link EsqlClientException},
+ * because then the error message sent back in the REST response would have an empty `type` string.
+ */
+public class DateMathException extends EsqlClientException {
```
@costin : As user-facing date math exceptions need to be thrown in more than 1 class now, I added this. This seemed to be the most pragmatic solution, but we can also resort to an anonymous class inheriting from EsqlClientException (with empty `type` attribute in the REST response, though), or make `EsqlClientException` non-abstract.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/AddTests.java:
```diff
@@ -99,7 +99,8 @@ public static Iterable<Object[]> parameters() {
                     new TestCaseSupplier.TypedData(rhs, EsqlDataTypes.DATE_PERIOD, "rhs")
                 ),
-                "AddDatetimesEvaluator[lhs=Attribute[channel=0], rhs=Attribute[channel=1]]",
+                // TODO: There is an evaluator for Datetime + Period, so it should be tested. Similarly below.
+                "No evaluator, the tests only trigger the folding code since Period is not representable",
```
@not-napoleon : This is a situation that I mentioned before, where it looks like we're testing the evaluator but we don't.

alex-spies committed:
Spotless

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/Neg.java:
```diff
@@ -67,29 +68,30 @@ else if (type == DataTypes.LONG) {
-            if (value instanceof Period period) {
-                return period.negated();
+        DataType dataType = field().dataType();
+        // For date periods and time durations, we need to handle folding especially. These types are unrepresentable, so there is no
```
```suggestion
        // For date periods and time durations, we need to treat folding differently. These types are unrepresentable, so there is no
```

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/Neg.java:
```diff
@@ -67,29 +68,30 @@ else if (type == DataTypes.LONG) {
+        // evaluator for them - but the default folding requires an evaluator.
+        if (dataType == DATE_PERIOD) {
+            Period fieldValue = (Period) field().fold();
+            try {
```
nit: maybe extract the try-catch into a method, to lower duplication?

alex-spies replied:
Tried that, annoyingly doesn't really seem to be possible.

`Period::negated` and `Duration::negated` are not implementations of a common interface, so I cannot abstract that away and need to branch based on the data type :/

bpintea commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/RestEsqlTestCase.java:
```diff
@@ -343,38 +343,37 @@ public void testErrorMessageForInvalidTypeInParams() throws IOException {
 
-    private static void assertDateMathOverflow(String overflowExpression, String expectedOverflowMessage) throws IOException {
+    private static void assertDateMathException(String overflowExpression, String errorSubstring) throws IOException {
         ResponseException re = expectThrows(
```
Will this always be a `DateMathException` in these tests?

alex-spies replied:
Good point, probably not; when the folding behavior evolves, I don't know what exactly we'll have there, maybe a more general `ExpressionException`, `InvalidQueryException` or so?

I can generalize the test to expect a `EsqlClientException` instead, this way future readers will not think that is _has_ to be exactly this type of exception (for whatever reason).

alex-spies replied:
Actually, I just realized that `assertDateMathException` does not really check for a `DateMathException`, making the name misleading. I'll update the name.

bpintea reviewed (approved):
LGTM

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateMathException.java:
```diff
@@ -0,0 +1,33 @@
+ * Even though this is a pretty specific exception, we cannot use e.g. a anonymous class that inherits from {@link EsqlClientException},
+ * because then the error message sent back in the REST response would have an empty `type` string.
+ */
+public class DateMathException extends EsqlClientException {
```
The users will not see the exception just the message thus there's no value in having a different exception type.
How many places are going to throw this error?
If there are multiple where the same error is thrown, it makes to centralize this into a class that creates the same EsqlClientException and return different messages.

alex-spies replied:
Unfortunately, this is not correct. The name of the exception is user-facing because of how the exception handling of our action listener framework works. When the REST response is being built in case of an error, the name of the exception class becomes the "type"; in the current state of this PR, the following query
```
row n = now() | eval n + (2147483647 year + 1 year)
```
results in the following (complete) response
```
{"error":{"root_cause":[{"type":"date_math_exception","reason":"arithmetic exception in expression [2147483647 year + 1 year]: [integer overflow]"}],"type":"date_math_exception","reason":"arithmetic exception in expression [2147483647 year + 1 year]: [integer overflow]"},"status":400}
```
Note the `"type": "date_math_exception"`. This was generated automatically, I don't have control over this.

Even though there's currently only 2 places where this exception is needed, it still kinda has to exist because the exception should inherit from `EsqlClientException` (abstract) and it should not be anonymous because of the reason above.

Since exceptions during constant folding is a general problem (see https://github.com/elastic/elasticsearch/issues/99758), I suspect that this exception can later be refactored into something more general, like `ConstantFoldingException`. I didn't use a more general name because names are hard and that discussion could be deferred until fold-time exceptions are handled consistently. But if you'd like me to create a more generic exception that inherits from `EsqlClientException`, instead, I'm happy to do so - maybe `ExpressionException`? (`VerificationException` comes to mind but is not really precise.)

costin replied:
Exception hierarchies in Java are tricky because they're abuse while they're suppose to have meaning.
It's the error message that is relevant, not the exception type. Since `EsqlClientException` is abstract you can use `ParsingException` or `VerificationException` - neither is really appropriate but will do for now.

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/NegTests.java:
```diff
@@ -139,32 +139,37 @@ public void testEdgeCases() {
 
-            return;
+            Duration minDuration = Duration.ofSeconds(Long.MIN_VALUE, 0);
+            try {
+                process(minDuration);
+            } catch (DateMathException e) {
+                assertEquals(
+                    e.getMessage(),
+                    "arithmetic exception in expression []: [Exceeds capacity of Duration: 9223372036854775808000000000]"
+                );
+                return;
+            }
+            throw new AssertionError("Expected exception when negating minimal time duration.");
```
use expectThrows() instead - see the other tests.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java:
```diff
@@ -133,7 +112,7 @@ public final Object fold() {
                 // Folding will be triggered before the plan is sent to the compute service, so we have to handle arithmetic exceptions
                 // manually and provide a user-friendly error message.
-                throw IllegalTemporalValueException.fromArithmeticException(source(), e);
+                throw DateMathException.fromArithmeticException(source(), e);
```
ExceptionUtils.math(ArithmeticException).

alex-spies replied:
I am not really sure how an `ExceptionUtils` class should look like to properly use it in this PR. As [mentioned above](https://github.com/elastic/elasticsearch/pull/99711#discussion_r1335691043), I need a new kind of `EsqlClientException`.

costin replied:
```
ExceptionUtils.math(Expression e, ArithmeticExpression ex) {
   return new ParsingException(e.source, "Expression [{}] causes arithmetic exception [{}], e, ex);
} 
```

alex-spies replied:
Alright, that should work!

costin reviewed (changes requested):
Drop the custom exception in favor of an ExceptionUtils class to use the same message/construct.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-date-math-for-negatives

alex-spies committed:
Factor out negation overflow test

alex-spies committed:
Improve comment wording

alex-spies committed:
Use better naming in test helper function

alex-spies committed:
Use expectThrows

costin reviewed (changes requested):
Drop the custom exception please.

alex-spies committed:
Remove DateMathException, add ExceptionUtils

Use VerificationException instead and add a static factory method in a
new ExceptionUtils class to construct this exception in case of invalid
date math expressions.

alex-spies commented:
@elasticsearchmachine run elasticsearch-ci/part-1

costin reviewed (approved):
LGTM
